### PR TITLE
Add eslint rule for preventing string literals in select/dispatch/useDispatch

### PR DESCRIPTION
**Pull request by david-szabo97**
## Description

Follow-up for #26655.

Add warning for these function calls: `select`, `dispatch`, `useDispatch`, `withDispatch`, `withSelect`, `createRegistrySelector`, `controls.select`, `controls.dispatch`, and `controls.resolveSelect` if the first argument is a literal.

Examples that will show a warning:
```js
select( 'core' )
dispatch( 'core ')
useDispatch( 'core' )
```

## How has this been tested?
1. Run `npm run lint-js`
2. Make sure it shows warnings for `select`, `dispatch`, `useDispatch`, `withDispatch`, `withSelect`, `createRegistrySelector`, `controls.select`, `controls.dispatch`, and `controls.resolveSelect`.

## Types of changes
Code quality

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
Let's downgrade those new rules to warning level and land immediately to simplify the process for remaining hardcoded store references.

**Comment by david-szabo97** [comment]
Found one false report. We use `select` in stories.

```js
import { text, select } from '@storybook/addon-knobs';
```

**Comment by gziolo** [comment]
> Found one false report. We use `select` in stories.
> 
> ```js
> import { text, select } from '@storybook/addon-knobs';
> ```

Yes, I see this being an issue in more places. It should match the following:
- `useDispatch`, `dispatch`, `select` and `resolveSelect` when imported from `@wordpress/data`
- `select` wrapped with `useSelect` call, `select` wrapped with `withSelect` and `dispatch` wrapper with `withDispatch`

Plus something you listed related to the registry maybe, definitely less common.

I don't mind we start small and expand later. It probably should always remind off in the recommended WordPress ESLint config since it's totally fine to reference a store as a string in a small plugin.

We should also document this rule and add tests (which is quite simple with the rule tester). Let me know how would you like to split work so I could help you land this PR.

**Comment by david-szabo97** [comment]
```
useDispatch, dispatch, select and resolveSelect when imported from @wordpress/data
select wrapped with useSelect call, select wrapped with withSelect and dispatch wrapper with withDispatch
```

I'm looking into how to make this more accurate. If you could take care of documenting and testing that would be great!

**Comment by david-szabo97** [comment]
The rule seems to work! Added tests as well.

File: .eslintrc.js [hunk 1]
```diff
@@ -61,6 +61,7 @@ module.exports = {
 			},
 		],
 		'@wordpress/no-unsafe-wp-apis': 'off',
+		'@wordpress/no-store-string-literals': 'warn',
```
Comment (gziolo) [inline .eslintrc.js:64]: It'd be great to prefix the rule name with "data" so it better explains its intent:  
```suggestion
		'@wordpress/data-no-store-string-literals': 'warn',
```

File: packages/eslint-plugin/rules/__tests__/no-store-string-literals.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * External dependencies
+ */
+import { RuleTester } from 'eslint';
+
+/**
+ * Internal dependencies
+ */
+import rule from '../no-store-string-literals';
+
+const ruleTester = new RuleTester( {
```
Comment (gziolo) [inline packages/eslint-plugin/rules/__tests__/no-store-string-literals.js:11]: This rule tester from ESLint is amazing. I love how easy it is to create all the checks. Great work here 👏🏻

File: packages/eslint-plugin/rules/no-store-string-literals.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+	return possibleCallExpressionNodes;
+}
+
+module.exports = {
+	meta: {
+		type: 'problem',
+		schema: [],
+		messages: {
+			doNotUseStringLiteral: `Do not use string literals ( '{{ argument }}' ) for accessing stores ; import the store and use the store object or store name constant instead`,
```
Comment (gziolo) [inline packages/eslint-plugin/rules/no-store-string-literals.js:93]: In the majority of cases, it should be the store definition so we could rephrase for simplicity to something like:
```suggestion
			doNotUseStringLiteral: `Do not use string literals ( '{{ argument }}' ) for accessing @wordpress/data stores. Pass the store definition instead`,
```

**Review by gziolo — APPROVED** [review]
I added missing documentation and left two minor comments. In general, this PR is ready to land. It just needs to be aligned in regards to the rule name. You can pick whatever works best for you but make sure you update docs accordingly.

Thank you for working on this rule. It's going to be super helpful ❤️

**Comment by gziolo** [comment]
All the issues reported by CI seem to be unrelated 🤷🏻‍♂️

**Pull request merged**